### ISSUE TITLE
Llama example uses ~10% of the CPU

### ISSUE THREAD

**Issue by kstavro** [huggingface/candle#1103]
When doing CPU inference with the llama example (x86, 16 threads), I notice only 10% CPU usage. This is in stark comparison to the quantized llama example, which makes 100% cpu usage during inference.

I guess this has to do with the fact that the ggml `matmul_q` kernels directly use `rayon` for the parallelism inside matmul, whereas the layers of the standard llama architecture use the external `gemm` implementations from `half` (if I am not mistaken). Which does have a `parallelism` input, so I thought this would lead to 100% cpu utilization.

Even though I have played around with BLAS in the past, I am far from an expert in it, so this might be working as intended? If not, I will post an issue in the gemm crate (as I think it really comes down to it after studying the kernels a bit), but I am first posting here to verify it is not just me and can be reproduced from others as well.

**Comment by LaurentMazare** [comment]
When using `gemm` we set the parallelism to the number of cores available on the local cpu [here](https://github.com/huggingface/candle/blob/588ad4835ae9a64f95c096feab8b96ff9238321d/candle-core/src/cpu_backend.rs#L1504) so should end up using all cores. Not sure how you get the % of usage but note that with `top` you should even see values greater than `100%` when using multiple cores.
You can try controlling the number of threads being used with the `RAYON_NUM_THREADS` environment variable to see how this affects the cpu usage as well as the number of token generated per second.

Finally on x86, you may want to use the mkl backend by installing the mkl and using `--features mkl` this may well result in faster inference and better cpu usage.

**Comment by okpatil4u** [comment]
I have observed this as well. Even with NUM_RAYON_THREADS setup, you do not
get proper CPU utilisation. With apple accelerate, most of the times CPU
usage gets capped at 125%.

**Comment by LaurentMazare** [comment]
The `accelerate` case is a bit different I think, it's likely that the neural engine of your mac gets used so you would see low CPU usage but still get faster performance than if your cpu was fully used (a bit similar than when running with cuda on a gpu on x86).

**Comment by okpatil4u** [comment]
Thanks Laurent. Does it mean that I won’t be able to use multiple threads
while using accelerate even when I am invoking them through different
applications ?

**Comment by LaurentMazare** [comment]
The neural engine behavior is undocumented so I have no clue how it handles multiple queries at the same time. Note that you can always deactivate accelerate if you prefer being sure that it runs on the CPU.

**Comment by okpatil4u** [comment]
That’s very interesting. I didn’t know that accelerate was running on ANE.
This explains the speed bottlenecks then.

**Comment by kstavro** [comment]
> When using `gemm` we set the parallelism to the number of cores available on the local cpu [here](https://github.com/huggingface/candle/blob/588ad4835ae9a64f95c096feab8b96ff9238321d/candle-core/src/cpu_backend.rs#L1504) so should end up using all cores. Not sure how you get the % of usage but note that with `top` you should even see values greater than `100%` when using multiple cores. You can try controlling the number of threads being used with the `RAYON_NUM_THREADS` environment variable to see how this affects the cpu usage as well as the number of token generated per second.
> 
> Finally on x86, you may want to use the mkl backend by installing the mkl and using `--features mkl` this may well result in faster inference and better cpu usage.

Thanks for the reply, This was tested on an AMD CPU with Windows, so the mkl feature is not available. CPU Utilization can be seen in the Task Manager on Windows instead of `top`. I have also printed the `parallelism` argument inside `gemm` just to be sure that all threads are seen and it is the full 16 threads, indeed. 

I know that there are a lot of parameters to be handled in the advanced `gemm` algorithms, like stack management etc which might have an influence on the amount of cores really needed for the computation? But on the other hand I am ignorant enough not to wonder if this 10% utilization might be something to look at. As far as I understand, the latest gemm kernels are from [this crate](https://github.com/sarah-ek/gemm/tree/main), right?

**Comment by LaurentMazare** [comment]
Yes the `gemm` crate is used under the hood. If you can trigger the issue using only a matrix multiplication in gemm that's super likely to be very helpful to them.

**Comment by kstavro** [comment]
Yep, I 'll try to reformulate this in a `gemm` fashion and take it there. Would keep this issue here open for a little more until the issue over at the gemm repo gets going, and then close it here, if that is alright with you.

**Comment by yportne13** [comment]
```rust
let k = Tensor::cat(&[k_cache, &k], 2)?.contiguous()?;
let v = Tensor::cat(&[v_cache, &v], 2)?.contiguous()?;
```
when kv_cache is long, this operation is too slow

**Comment by kstavro** [comment]
Small update from my side. So I found some time yesterday and got to the bottom of it: the **tl;dr** is that only `gemm` is really parallelized, whereas `gevv` and `gemv` are not. According to the author of the gemm crate, both `gevv` and `gemv` are memory-bound, do there is no benefit from parallelizing.

What now happens during inference is that `kv_cache` kicks in, so all the matmuls there are `gemv` because `m=1` due to the nature of how `kv_cache` works (there are some dot products as well where `k=1` and `gevv` kicks in). So this explains it.

Being quite ignorant to the intricacies of SIMD etc, I still find it sort of surprising that at least for `gemv` it makes no sense to make it parallel. Most probably the author of the crate has already tried formulating gemv as a parallel gevvs and that probably brings nothing, so what do I know. Just noting that, as far as I understand, `QMatmul`s inside candle are practically parallel SIMD gevvs, and that seems to work quite well.

For more information, you can have a look at the linked issue over at gemm. I might still play a bit with gemv vs parallel gevvs out of curiosity, but I wouldn't expect much coming out of it.
